grpc-protocol/src/main/java/com/linecorp/armeria/client/grpc/protocol/UnaryGrpcClient.java

```diff
@@ -239,33 +289,33 @@ public void onNext(DeframedMessage message) {
                         return;
                     }
                     if (processedMessages > 0) {
-                        onError(new ArmeriaStatusException(StatusCodes.INTERNAL,
-                                                           "received more than one data message; " +
-                                                           "UnaryGrpcClient does not support streaming."));
+                        cancel(new ArmeriaStatusException(StatusCodes.INTERNAL,
+                                                          "received more than one data message; " +
+                                                          "UnaryGrpcClient does not support streaming."));
                         return;
                     }
                     final ByteBuf buf = message.buf();
                     // Compression not supported.
                     assert buf != null;
                     content = HttpData.wrap(buf);
+                    buf.retain();
```

Comment:
How about managing the reference count in this `process` method instead of releasing it `onNext`: https://github.com/line/armeria/pull/3736/files#diff-4798ea4651999fb5f62ea8b9576d6603af31de5f74e996cff16fcc5a62204624R297 so that we don't have to call `retain`?

Reply:
It's done this way to clearly separate resource management for the `DeframedMessage` and the processing logic. `onNext` is a final consumer of the `DeframedMessage` and is responsible for closing it to avoid leaks, regardless of the business logic inside `process`. That's why `message.close` is happening in the `finally` statement there. `process` is now only responsible for retaining the buffer in case it needs to keep it inside the `HttpData` to return in the response (only one case our of multiple branches), since `buf` is explicitly retained in that case - even if someone rearranges the code or adds more flow control they don't need to worrying about releasing the buffer to avoid memory leak because `onNext` takes care of that.

Reply:
Thanks for the elaboration. SGTM.